### PR TITLE
WIP - corrected & improved Filter and Reverse

### PR DESCRIPTION
**Pull request by shawnce**
@gribozavr FYI

### CODE REVIEW

File: stdlib/public/core/Character.swift [hunk 1]
```diff
@@ -93,7 +93,7 @@ public struct Character :
     _precondition(
       s._core.count != 0, "Can't form a Character from an empty String")
     _precondition(
-      s.startIndex.successor() == s.endIndex,
+      s.characters.next(s.startIndex) == s.endIndex,
```
Comment (shawnce) [inline stdlib/public/core/Character.swift]: I guess we should expose next(Index), etc. on String to make the API symmetric with having startIndex and endIndex?

Comment (gribozavr) [inline stdlib/public/core/Character.swift]: Yes, that would be in line with the current definition of String.

File: stdlib/public/core/Filter.swift [hunk 2]
```diff
@@ -222,9 +230,13 @@ public struct LazyFilterCollection<
...
   public func advance(i: Index, by n: IndexDistance, limit: Index) -> Index {
+    _precondition(n >= 0,
+      "Only BidirectionalCollections can be advanced by a negative amount")
+    // TODO: swift-3-indexing-model: _failEarlyRangeCheck i?
+
     var index = i.base
-    for _ in 0..<n where index != _base.endIndex && index != limit.base {
-      index = _nextFiltered(index)
+    for _ in 0..<n {
+      _nextFilteredInPlace(&index, limit: limit.base)
```
Comment (gribozavr) [inline stdlib/public/core/Filter.swift]: Once we hit the limit, we should stop looping.  `_nextFilteredInPlace` has that information, it should probably return it.

File: stdlib/public/core/StringUTF16.swift [hunk 3]
```diff
@@ -45,24 +47,28 @@ extension String {
     // TODO: swift-3-indexing-model - add docs
     @warn_unused_result
     public func next(i: Index) -> Index {
+      // FIXME: swift-3-indexing-model: range check i?
       return Index(_offset: i._offset.advanced(by: 1))
```
Comment (gribozavr) [inline stdlib/public/core/StringUTF16.swift]: Please use the `_unsafePlus` and `_unsafeMinus` instead.  They avoid overflow checks.  We should be doing bounds checks instead, like the fixme says.

**Comment by gribozavr** [comment]
@shawnce LGTM, thanks!

File: stdlib/public/core/Filter.swift (hunk 2 above)
Comment (shawnce) [inline stdlib/public/core/Filter.swift]: Wow... left my last round of changes stashed it looks like, apply now and will submit again

File: stdlib/public/core/StringUTF16.swift (hunk 3 above)
Comment (shawnce) [inline stdlib/public/core/StringUTF16.swift]: Yeah I wasn't thinking far enough that the Index was actually an Int... will update.

Comment (shawnce) [inline stdlib/public/core/StringUTF16.swift]: @gribozavr 

I am wondering what you are thinking for bounds checking of an index when advancing it in some way especially when Indexes are Int (or Int like). For example something like the following with a pre and post condition check? What to do about the boundary conditions with overflow and underflow?

```
    @warn_unused_result
    public func next(i: Index) -> Index {
      _rangeCheck(i._offset, bounds: 0..<_length)
      let offset = _unsafePlus(i._offset, 1)
      _rangeCheck(offset, bounds: 0..<_length)
      return Index(_offset: offset)
    }
    public func _rangeCheck(index: int, bounds: Range<Int>) {
      _precondition(bounds.startIndex <= index, "blah")
      _precondition(index < bounds.endIndex, "blah")
    }
```

(ignoring code reuse across advance, previous, etc)

Comment (gribozavr) [inline stdlib/public/core/StringUTF16.swift]: This is addressed in https://github.com/apple/swift/pull/1592

**Comment by gribozavr** [comment]
All review comments are addressed in https://github.com/apple/swift/pull/1592.

**Pull request merged**